packages/devtools-map-bindings/src/parser.js

```diff
@@ -0,0 +1,297 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+
+import type { Source, SourceId, Location, SourceScope } from "debugger-html";
+import type { NodePath, Node, Location as BabelLocation } from "babel-traverse";
+
+export type ParsedScopeNamesReferences = {
+  [name: string]: BabelLocation[]
+};
+
+export type ParsedScope = {
+  start: Location,
+  end: Location,
+  type: string,
+  bindings: ParsedScopeNamesReferences,
+  children: ?(ParsedScope[])
+};
+
+export type ParseJSScopeVisitor = {
+  traverseVisitor: any,
+  toParsedScopes: () => ParsedScope[],
+};
+
+type TempScopeNameReferences = {
+  type: string,
+  refs: BabelLocation[]
+};
+
+type TempScopeNamesReferences = {
+  [name: string]: TempScopeNameReferences
+};
+
+type TempScope = {
+  type: string,
+  parent: TempScope | null,
+  children: Array<TempScope>,
+  loc: BabelLocation,
+  names: TempScopeNamesReferences
+};
+
+function createTempScope(
+  type: string,
+  parent: TempScope | null,
+  loc: BabelLocation
+): TempScope {
+  const result = {
+    type,
+    parent,
+    children: [],
+    loc: loc,
+    names: Object.create(null)
+  };
+  if (parent) {
+    parent.children.push(result);
+  }
+  return result;
+}
+
+function isNode(node?: Node, type: string): boolean {
+  return node ? node.type === type : false;
+}
+
+function getFunctionScope(scope: TempScope): TempScope {
+  let s = scope;
+  while (s.type !== "Function" && s.type !== "Script") {
+    if (!s.parent) {
+      return s;
+    }
+    s = s.parent;
+  }
+  return s;
+}
+
+function fromBabelLocation(
+  location: BabelLocation,
+  sourceId: SourceId
+): Location {
+  return {
+    sourceId,
+    line: location.line,
+    column: location.column
+  };
+}
+
+function parseDeclarator(
+  declaratorId: Node,
+  targetScope: TempScope,
+  type: string
+) {
+  if (isNode(declaratorId, "Identifier")) {
+    targetScope.names[declaratorId.name] = { type, refs: [] };
+    return;
+  } else if (isNode(declaratorId, "ObjectPattern")) {
+    declaratorId.properties.forEach(prop => {
+      parseDeclarator(prop.key, targetScope, type);
+    });
+  } else if (isNode(declaratorId, "ArrayPattern")) {
+    declaratorId.elements.forEach(item => {
+      parseDeclarator(item, targetScope, type);
+    });
+  } else if (isNode(declaratorId, "RestElement")) {
+    parseDeclarator(declaratorId.argument, targetScope, type);
+  }
+}
+
+function recordIdentifierPosition(
+  scope: TempScope,
+  name: string,
+  loc: BabelLocation
+) {
+  // Find nearest outer scope with the specifed name and add reference.
+  for (let s = scope; s; s = s.parent) {
+    if (name in s.names) {
+      s.names[name].refs.push(loc.start);
```

Comment:
it would be nice if this function avoided mutating `scope` and instead returned the found location: